npeps/npep-133-fqdn-egress-selector.md

```diff
@@ -0,0 +1,438 @@
+spec:
+  priority: 55
+  subject:
+    namespaces:
+      matchLabels:
+        kubernetes.io/metadata.name: "monitoring"
+  egress:
+  - name: "allow-to-my-service"
+    action: "Allow"
+    to:
+    - fqdn:
+        name: "my-service.com"
+    - fqdn:
+        pattern: "*.cloud-provider.io"
+    ports:
+    - portNumber:
+        protocol: TCP
+        port: 443
+  - name: "default-deny"
+    action: "Deny"
+    to:
+    - networks:
+      - "0.0.0.0/0"
+```
+
+This example uses a default-deny BaselineAdminNetworkPolicy to create the
+allowlist:
+```yaml
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: AdminNetworkPolicy
+metadata:
+  name: allow-my-service-egress
+spec:
+  priority: 55
+  subject:
+    namespaces:
+      matchLabels:
+        kubernetes.io/metadata.name: "monitoring"
+  egress:
+  - name: "allow-to-my-service"
+    action: "Allow"
+    to:
+    - fqdn:
+        name: "my-service.com"
+    - fqdn:
+        pattern: "*.cloud-provider.io"
+    ports:
+    - portNumber:
+        protocol: TCP
+        port: 443
+---
+apiVersion: policy.networking.k8s.io/v1alpha1
+kind: BaselineAdminNetworkPolicy
+metadata:
+  name: default
+spec:
+  subject:
+    namespaces: {}
+  ingress:
+    - action: Deny
+      to:
+      - networks:
+        - "0.0.0.0/0"
+```
+
+### Expected Behavior
+
+1. A FQDN egress policy does not grant the workload permission to communicate
+   with any in-cluster DNS services (like `kube-dns`). A separate rule needs to
+   be configured to allow traffic to any DNS servers.
+1. FQDN policies should not affect the ability of workloads to resolve domains,
+   only their ability to communicate with the IP backing them.
+   *  For example, if a policy allows traffic to `kubernetes.io`, any selected
+      Pods can still resolve `wikipedia.org` or
+      `my-services.default.svc.cluster.local`, but can not send traffic to them
+      unless allowed by a different rule.
+1. Each implementation will provide guidance on which DNS name-server is
+   considered authoritative for resolving domain names. This could be the
+   `kube-dns` Service or potentially some other DNS provider specified in the
+   implementation's configuration.
+1. DNS record querying and lifetimes:
+   *  Pods are expected to make a DNS query for a domain before sending traffic
+      to it. If the Pod fails to send a DNS request and instead just sends
+      traffic to the IP (either because of caching or a static config), traffic
+      is not guaranteed to flow.
+   *  Pods should respect the TTL of DNS records they receive. Trying to
+      establish new connection using DNS records that are expired is not
+      guaranteed to work.
+   *  When the TTL for a DNS record expires, the implementor should stop
+      allowing new connections to that IP. Existing connection will still be
+      allowed (that's consistent with NetworkPolicy behavior on long-running
+      connections). 
+1. Targeting in-cluster endpoints with FQDN selector:
```

Comment:
Wait in the "non-goals" section above don't we say 

```
Using the FQDN selector to refer to other Kubernetes endpoints, while not
    explicitly disallowed, is not defined by this spec and left up to individual
    providers. Trying to allow traffic to the following domains is NOT
    guaranteed to work:
    * `my-svc.my-namespace.svc.cluster.local` (the generated DNS record for a
      Service as defined
      [here](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#services))
    * `pod-ip-address.my-namespace.pod.cluster.local` (the generated DNS record
      for a Pod as defined
      [here](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pods))
``` 

So why is this "supported' here?

Reply:
Whoops, good catch. This was the way I originally wanted to frame it. 
But given the way we defined the network API, I think the discussion in "expected behavior" is more correct.

I'm just removing the section from non-goals.